src/autopas/pairwiseFunctors/LJFunctorCuda.cu

```diff
@@ -0,0 +1,907 @@
+/**
+ * @file LJFunctorCuda.h
+ *
+ * @date 9.1.2019
+ * @author jspahl
+ */
+#include "autopas/pairwiseFunctors/LJFunctorCuda.cuh"
+#include <iostream>
+#include "autopas/utils/ExceptionHandler.h"
+#include "autopas/utils/CudaExceptionHandler.h"
+#include "math_constants.h"
+
+namespace autopas {
+
+__constant__ constants<float> global_constants_float;
+__constant__ constants<double> global_constants_double;
+
+template<typename floatType>
+class soa {
+public:
+	floatType* posX;
+	floatType* posY;
+	floatType* posZ;
+	floatType* forceX;
+	floatType* forceY;
+	floatType* forceZ;
+};
+
+template<typename T>
+__device__ inline constants<T>& getConstants() {
+	return global_constants_float;
+}
+template<>
+__device__ inline constants<double>& getConstants<double>() {
+	return global_constants_double;
+}
+
+template<typename T>
+__device__ inline T getInfinity() {
+	return CUDART_INF_F;
+}
+template<>
+__device__ inline double getInfinity<double>() {
+	return CUDART_INF;
+}
+
+template<typename floatType>
+__device__
+inline typename vec3<floatType>::Type bodyBodyF(
+		typename vec3<floatType>::Type i, typename vec3<floatType>::Type j,
+		typename vec3<floatType>::Type fi) {
+	floatType drx = i.x - j.x;
+	floatType dry = i.y - j.y;
+	floatType drz = i.z - j.z;
+
+	floatType dr2 = drx * drx + dry * dry + drz * drz;
+
+	if (dr2 > getConstants<floatType>().cutoffsquare | dr2 == 0.0) {
+		return fi;
+	}
+
+	floatType invdr2 = 1. / dr2;
+	floatType lj6 = getConstants<floatType>().sigmasquare * invdr2;
+	lj6 = lj6 * lj6 * lj6;
+	floatType lj12 = lj6 * lj6;
+	floatType lj12m6 = lj12 - lj6;
+	floatType fac = getConstants<floatType>().epsilon24 * (lj12 + lj12m6)
+			* invdr2;
+
+	fi.x += drx * fac;
+	fi.y += dry * fac;
+	fi.z += drz * fac;
+
+	return fi;
+}
+
+template<typename floatType, bool n3AdditionSafe = false>
+__device__
+inline typename vec3<floatType>::Type bodyBodyFN3(
+		typename vec3<floatType>::Type i, typename vec3<floatType>::Type j,
+		typename vec3<floatType>::Type fi, typename vec3<floatType>::Type* fj) {
+	floatType drx = i.x - j.x;
+	floatType dry = i.y - j.y;
+	floatType drz = i.z - j.z;
+
+	floatType dr2 = drx * drx + dry * dry + drz * drz;
+
+	if (dr2 > getConstants<floatType>().cutoffsquare) {
+		return fi;
+	}
+
+	floatType invdr2 = 1. / dr2;
+	floatType lj6 = getConstants<floatType>().sigmasquare * invdr2;
+	lj6 = lj6 * lj6 * lj6;
+	floatType lj12 = lj6 * lj6;
+	floatType lj12m6 = lj12 - lj6;
+	floatType fac = getConstants<floatType>().epsilon24 * (lj12 + lj12m6)
+			* invdr2;
+
+	floatType dfx = drx * fac;
+	floatType dfy = dry * fac;
+	floatType dfz = drz * fac;
+
+	fi.x += dfx;
+	fi.y += dfy;
+	fi.z += dfz;
+
+	if (n3AdditionSafe) {
+		fj->x -= dfx;
```

Comment:
this is only allowed to be called safely if no conflicts can occur. How is this guaranteed?

Reply:
?